### PR TITLE
lane colorization and color smoother

### PR DESCRIPTION
**Pull request by xjh19971**
Add a option to create colored lane map when running play_map.py.  Use color_smoother.py to post-process lane map. Add bezier package into environment.yaml. Due to the compatibility of bezier, python must be downgrade to 3.6.

### CODE REVIEW

File: environment.yaml [hunk 1]
```diff
@@ -5,7 +5,7 @@ channels:
 dependencies:
   - imageio
   - pandas
-  - python=3.8
+  - python=3.6
```
Comment (Atcold) [inline environment.yaml:8]: Plz, revert back to `3.8`.

File: traffic_gym.py [hunk 2]
```diff
@@ -671,6 +674,9 @@ def __init__(self, display=True, nb_lanes=4, fps=30, delta_t=None, traffic_rate=
         self.user_is_done = None
 
         self.display = display
+
+        self.trajectory_image = None
+        self.observed_car = []
```
Comment (Atcold) [inline traffic_gym.py:679]: Use `list()` instead.

File: traffic_gym.py [hunk 3]
```diff
@@ -924,15 +932,23 @@ def render(self, mode='human', width_height=None, scale=1.):
                     self.screen.blit(self.photos[i], self.photos_rect[i])
 
             # draw lanes
-            self._draw_lanes(self.screen)
+            self._draw_lanes(self.screen,colored_lane=self.colored_lane)
```
Comment (Atcold) [inline traffic_gym.py:935]: Space missing.

File: environment.yaml (hunk 1 above)
Comment (xjh19971) [reply environment.yaml]: Env.yaml is changed to python 3.8. color_smoother.py adds EOL.

File: traffic_gym.py (hunk 2 above)
Comment (xjh19971) [reply traffic_gym.py]: list() is used.

File: traffic_gym.py (hunk 3 above)
Comment (xjh19971) [reply traffic_gym.py]: Space is added.

**Comment by xjh19971** [comment]
I revise some errors and add an option to use colored map as background for generate_data and train_fm. Please review.

File: map_i80.py [hunk 4]
```diff
@@ -483,76 +483,85 @@ def step(self, policy_action=None):
...
+                ])
+                pygame.Surface.fill(
+                    s,
+                    b,
+                    pygame.Rect(m + 60 * LANE_W, bottom + 5, sw - 60 * LANE_W, 54 + 30 - 5))
+
+                self._lane_surfaces[mode] = surface.copy()
+            else:
+                s = surface  # screen
+                lane = pygame.image.load(colored_lane)
```
Comment (Atcold) [inline map_i80.py:565]: This is wrong.
The `'machine'` view has an `m` offset. Did you try saving the `Surface` you're getting?

File: traffic_gym.py [hunk 5]
```diff
@@ -1024,6 +1040,61 @@ def _draw_lanes(self, surface, mode='human', offset=0):
...
+                                                                                  axis=2)
+        hsv = rgb_to_hsv(self.trajectory_image)
+        s = hsv[:, :, 1]
+        s[s < threshold] = 0
+        hsv[:, :, 1] = s
+        v = hsv[:, :, 2]
+        v[v > 0] = 1
+        hsv[:, :, 2] = v
+        self.trajectory_image = hsv_to_rgb(hsv)
+        plt.imsave("1trajectory.jpg", np.transpose(self.trajectory_image, (1, 0, 2)))
```
Comment (Atcold) [inline traffic_gym.py:1096]: Is the `1` a typo?

Comment (xjh19971) [reply traffic_gym.py:1096]: It is a symbol for original lane map. I will remove it.

File: map_i80.py (hunk 4 above)
Comment (xjh19971) [reply map_i80.py:567]: i80-machine
i80-machine-1

I add offset for machine mode. Seems works well. Should I add blue background for colored map?

**Comment by xjh19971** [comment]
vehicle_surface

Vehicles is covered by colored map. Need to separate a vehicle map.

**Comment by xjh19971** [comment]
Temporaily close this PR until forward model can be used to train policy.

File: map_i80.py (hunk 4 above)
Comment (Atcold) [reply map_i80.py:567]: I think not. The rails are going to be attractors. Before, the red and blue were repellers.

**Comment by xjh19971** [comment]
A basic version for colored lane usage has been finished.

**Comment by xjh19971** [comment]
More ideas. Temporarily close this PR to implement them.

**Comment by Atcold** [comment]
Hehe, okay 😄

**Pull request closed without merging**